portality/scripts/application_status_report.py

```diff
@@ -59,27 +72,48 @@ def status_query(date_year, status):
 
     parser = argparse.ArgumentParser()
     parser.add_argument("-o", "--out", help="output file path")
-    parser.add_argument("-y", "--year", help="year to filter by")
+    parser.add_argument("-fy", "--from_year", help="from year to filter by")
+    parser.add_argument("-ty", "--to_year", help="to year to filter by")
     args = parser.parse_args()
 
     if not args.out:
         print("Please specify an output file path with the -o option")
-        parser.print_help()
-        exit()
 
-    if not args.year:
-        print("Please specify a year to filter the applications with the -y option")
+    if not args.from_year:
+        print("Please specify a 'from year' to filter the applications with the -fy option")
+
+    if not args.to_year:
+        print("Please specify a 'to year' to filter the applications with the -ty option")
+
+    if not args.out or not args.from_year or not args.to_year:
         parser.print_help()
         exit()
 
     with open(args.out, "w", encoding="utf-8") as f:
         writer = csv.writer(f)
 
-        res = models.Application.query(q=date_applied_query(args.year), size=0)
+        writer.writerow(["Applications data years", args.from_year + "-" + args.to_year])
+        writer.writerow([])
+
+        res = models.Application.query(q=date_applied_query(args.from_year, args.to_year), size=0)
         writer.writerow(["Submitted", res.get("hits", {}).get("total", {}).get("value", 0)])
 
-        res = models.Provenance.query(q=status_query(args.year, "accepted"), size=0)
-        writer.writerow(["Accepted", res.get("hits", {}).get("total", {}).get("value", 0)])
+        accepted_counter = 0
+        rejected_counter = 0
+
+        # Application id is saved as resource_id in Provenance model
+        # Iterate through all the applications and query Provenance with application id and status to retrieve the
+        # status of the applications
+        for record in models.Application.iterate(q=date_applied_query(args.from_year, args.to_year), page_size=20):
```

Comment:
hmm, this might be an inefficient use of the index - we're querying Applications in a date range with size 0 (fine) then we're iterating through all them (a second query), then we're performing 2 Provenance queries per application.

And here we miss an application that was submitted in one year and changed state in the next, which I think might happen frequently. That suggests we should remove the date range from the provenance queries entirely i.e.:

**for application X that applied in $year, was it ultimately accepted or rejected?**

On the other hand, this isn't how the task is worded, so we should keep our current assumptions, i.e:

* **How many applications were submitted (`created_date` in $year)**
* **how many applications were accepted / rejected (`Provenance` records in $year)**

In which case, I think we need to look through the Provenance index for any accepted or rejected, then look up whether that's an Application or not. What you currently have is halfway between the two interpretations:

**count of applications submitted in $year that were also accepted / rejected in $year**

I think there's a case for clarifying which interpretation of the task to use, but for now we should commit to one of them. How about iterating to get all IDs of applications, then for each accepted / rejected `Provenance` record in $year check it's a new application? (ID appears in our list) That reduces our index to 2n scrolls, rather than n^2 queries and a scroll.

Reply:
Here's my take on what the overall effect of the query should be:

1. Get each application (not UR) submitted prior to the start of 2019
2. Record the year that the application was submitted (the created_date)
3. Get the rejected or accepted provenance record for the application
4. Record the year and whether it was rejected or accepted

In terms of efficient implementation, I'm not immediately sure what the best is, but your options are:

1. Scroll the applications limited by the query in (1) above, then query the Provenance for that app id and the status accepted or rejected, which should give you only 1 result (1 scroll, n additional queries)
2. Scroll the applications by the query in (1), and record ALL the application ids and created_dates in memory.  Then scroll the provenance records for rejected or accepted applications and pick out the ids and the status from those that are already in memory (2 scrolls)

Reply:
@Steven-Eardley 

> hmm, this might be an inefficient use of the index - we're querying Applications in a date range with size 0 (fine) then we're iterating through all them (a second query), then we're performing 2 Provenance queries per application.

Yes, agree I could have thought better implementation

> How many applications were submitted (created_date in $year)
how many applications were accepted / rejected (Provenance records in $year)

Yes, I understood the requirement this way

> In which case, I think we need to look through the Provenance index for any accepted or rejected, then look up whether that's an Application or not. What you currently have is halfway between the two interpretations:

I don't think we can find if the Provenance record is an application or UR. The type field in the record is recorded as 'suggestion' in both the cases. We can find the application record only through the application id from resource id. Incase of UR, resource_id is not application id. It is new unique id for each record incase of UR.

@richard-jones 
To me option (1) is the better option.

> Scroll the applications limited by the query in (1) above, then query the Provenance for that app id and the status accepted or rejected, which should give you only 1 result (1 scroll, n additional queries)

Here there can be a chance of more than one record. For example if the application got rejected and then accepted and vice versa. In this case I think we can pull latest record and get the status